File: _IFSP/LP2/Exercicios/Jogos/Classes/hanoi.py
class HanoiTower:
    def __init__(self) -> None:
        self.tower = [[1, 0, 0], [2, 0, 0], [3, 0, 0], [4, 0, 0], [5, 0, 0]]

    def win_check(self) -> bool:
        return all(self.tower[i][2] != 0 for i in range(5))

    def print_hanoi(self) -> None:
        for row in self.tower:
            print(" ".join(map(str, row)))
        print()

    def validate(self, init:int, destiny:int) -> bool:
        return not (init == destiny or init < 1 or init > 3 or destiny < 1 or destiny > 3 or self.tower[4][init - 1] == 0)

    def get_top(self, column:int) -> int:
        for i in range(5):
            if self.tower[i][column] != 0:
                return i
        return 5

    def swap_hanoi(self, init:int, destiny:int) -> bool:
        init -= 1
        destiny -= 1
        idx_top_init = self.get_top(init)
        idx_top_destiny = self.get_top(destiny)

        if idx_top_destiny < 5 and self.tower[idx_top_init][init] > self.tower[idx_top_destiny][destiny]:
            return False

        if idx_top_destiny == 5:
            self.tower[idx_top_init][init], self.tower[4][destiny] = self.tower[4][destiny], self.tower[idx_top_init][init]
        else:
            self.tower[idx_top_init][init], self.tower[idx_top_destiny - 1][destiny] = self.tower[idx_top_destiny - 1][destiny], self.tower[idx_top_init][init]
       
        return True


def main():
    game = HanoiTower()

    while True:
        game.print_hanoi()
        init, destiny = map(int, input("Escolha seu movimento [inicio | destino](1|2|3): ").split())

        if game.validate(init, destiny):
            if game.swap_hanoi(init, destiny):
                print("ok!")
            else:
                print("movimento invalido")

            if game.win_check():
                game.print_hanoi()
                print("voce ganhou!\n")
                break
        else:
            print("movimento invalido")


main()

File: 01_IFSP/LP2/Exercicios/Jogos/Classes/hanoi.c
#include <stdio.h>

int hanoiTower[5][3] = {{1, 0, 0}, {2, 0, 0}, {3, 0, 0}, {4, 0, 0}, {5, 0, 0}};

void swap(int *x, int *y)
{
    int temp;
    temp = *x;
    *x = *y;
    *y = temp;
}

int winCheck()
{
    int ant = 0;
    for (int i = 0; i < 5; i++)
    {
        if (hanoiTower[i][2] <= ant)
        {
            return 0;
        }
        ant = hanoiTower[i][2];
    }
    return 1;
}

void printHanoi()
{
    for (int i = 0; i < 5; i++)
    {
        for (int j = 0; j < 3; j++)
        {
            printf("%d ", hanoiTower[i][j]);
        }
        printf("\n");
    }
    printf("\n");
}

int validate(int init, int destiny)
{
    return !(init == destiny || init < 1 || init > 3 || destiny < 1 || destiny > 3 || hanoiTower[4][init - 1] == 0);
}

int getTop(int column)
{
    for (int i = 0; i < 5; i++)
    {
        if (hanoiTower[i][column] != 0)
            return i;
    }
    return 5;
}

int swapHanoi(int init, int destiny)
{
    init--;
    destiny--;
    int idxTopInit = getTop(init);
    int idxTopDestiny = getTop(destiny);
    if (hanoiTower[idxTopDestiny][destiny] != 0 && hanoiTower[idxTopInit][init] > hanoiTower[idxTopDestiny][destiny])
    {
        return 0;
    }
    if (idxTopDestiny == 5)
    {
        swap(&hanoiTower[idxTopInit][init], &hanoiTower[4][destiny]);
        return 1;
    }
    swap(&hanoiTower[idxTopInit][init], &hanoiTower[idxTopDestiny - 1][destiny]);
    return 1;
}

int main()
{
    int init, destiny;

    while (1)
    {
        printHanoi();
        printf("escolha seu movimento [inicio | destino](1|2|3): ");
        scanf("%d%d", &init, &destiny);

        if (validate(init, destiny))
        {

            if (swapHanoi(init, destiny))
            {
                printf("ok!\n");
            }
            else
            {
                printf("movimento invalido\n");
            }
            if (winCheck())
            {
                printHanoi();
                printf("voce ganhou!\n\n");
                break;
            }
        }
    }

    return 0;
}
